File: 1769-minimum-number-of-operations-to-move-all-balls-to-each-box/1769-minimum-number-of-operations-to-move-all-balls-to-each-box.py
class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        n = len(boxes)
        go_pre = [0] # go_pre[i] = i-1 까지의 1 갯수
        back_pre = [0]
        for i in range(n):
            go_pre.append(go_pre[-1] + (boxes[i] == '1'))
            back_pre.append(back_pre[-1] + (boxes[n - i - 1] == '1'))

        go_dp = [0] * n
        back_dp = [0] * n
        for i in range(1, n):
            go_dp[i] = go_dp[i - 1] + go_pre[i]
            back_dp[i] = back_dp[i - 1] + back_pre[i]

        return [go_dp[i] + back_dp[n-i-1] for i in range(n)]

File: 1769-minimum-number-of-operations-to-move-all-balls-to-each-box/1769-minimum-number-of-operations-to-move-all-balls-to-each-box.rs
impl Solution {
    pub fn min_operations(boxes: String) -> Vec<i32> {
        let boxes: Vec<i32> = boxes.chars().map(|x| x as i32 - '0' as i32).collect();
        let n: usize = boxes.len();
        let mut res: Vec<i32> = vec![0; n];

        let mut move_left = 0; // 현재 값
        let mut ball_left = 0; // 하나 이전 보관용
        let mut move_right = 0;
        let mut ball_right = 0;

        for i in 0..n {
            res[i] += move_left;
            ball_left += boxes[i];
            move_left += ball_left;

            res[n-i-1] += move_right;
            ball_right += boxes[n-i-1];
            move_right += ball_right;
        }

        res
    }
}